File: 1주차/mijip0320/string_compression.js
var compresss = function (chars) {
    
    
    let write = 0; //placing the compressed characters in the list 
    let read = 0; //iterating through the list.

    while (read < chars.length) {
        let char = chars[read];
        let count = 0;

        // Count the number of consecutive characters
        while (read < chars.length && chars[read] === char) {
            count++;
            read++;
        }

        //Replace the character at the write position in the list with the current character.
        chars[write] = char;
        //move pointer
        write++;
        if (count > 1) {
            //If the character repeats more than once, convert the count to a string and add each digit to the list at subsequent positions.
            for (let digit of String(count)) {
                chars[write] = digit;
                write++;
            }
        }
    }

    //return write;

    console.log(chars)
};

File: 1주차/desiresprout/string compression.py
# 1. Problem 

# 2. Test case
# 	Input: the sky is blue
# 	Output: blue is sky the
			
# 	Input: "  hello world  "
# 	Output: world hello
		
	
# 3. Plan
# Pointer where we write the compressed string
# Pointer for reading through chars array
# Count the occurrences of the current character
# Write the character to the `write` pointer
# If the count is more than 1, write the count as well
	
	
# Time Complexity: O(n)
# Space Complexity: O(1)

def compress(chars):
    write = 0  
    read = 0 
    
    while read < len(chars):
        char = chars[read]  
        count = 0
        
       
        while read < len(chars) and chars[read] == char:
            read += 1
            count += 1
        
        chars[write] = char
        write += 1
        
        if count > 1:
            for digit in str(count):
                chars[write] = digit
                write += 1
    
    return write


